src/catalog/src/durable/persist.rs

```diff
@@ -984,30 +1018,39 @@ impl UnopenedPersistCatalogState {
         self.mode = mode;
         let read_only = matches!(self.mode, Mode::Readonly);
 
-        self.sync_to_current_upper().await?;
-        let prev_epoch = self.epoch.validate()?;
         // Fence out previous catalogs.
-        let mut fence_updates = Vec::with_capacity(2);
-        if let Some(prev_epoch) = prev_epoch {
-            fence_updates.push((StateUpdateKind::Epoch(prev_epoch), -1));
-        }
-        let mut current_epoch = prev_epoch.unwrap_or(MIN_EPOCH).get();
-        // Only writable catalogs attempt to increment the epoch.
-        if matches!(self.mode, Mode::Writable) {
-            current_epoch = current_epoch + 1;
-        }
-        let current_epoch = Epoch::new(current_epoch).expect("known to be non-zero");
-        fence_updates.push((StateUpdateKind::Epoch(current_epoch), 1));
-        let current_epoch = FenceableToken::new(Some(current_epoch), FenceError::epoch);
-        debug!(
-            ?self.upper,
-            ?prev_epoch,
-            ?current_epoch,
-            "fencing previous catalogs"
-        );
-        self.epoch = current_epoch;
-        if matches!(self.mode, Mode::Writable) {
-            self.compare_and_append(fence_updates).await?;
+        loop {
+            self.sync_to_current_upper().await?;
+            let prev_epoch = self.epoch.validate()?;
+            let mut fence_updates = Vec::with_capacity(2);
+            if let Some(prev_epoch) = prev_epoch {
+                fence_updates.push((StateUpdateKind::Epoch(prev_epoch), -1));
+            }
+            let mut current_epoch = prev_epoch.unwrap_or(MIN_EPOCH).get();
+            // Only writable catalogs attempt to increment the epoch.
+            if matches!(self.mode, Mode::Writable) {
+                current_epoch = current_epoch + 1;
+            }
+            let current_epoch = Epoch::new(current_epoch).expect("known to be non-zero");
+            fence_updates.push((StateUpdateKind::Epoch(current_epoch), 1));
+            let current_epoch = FenceableToken::new(Some(current_epoch), FenceError::epoch);
+            debug!(
+                ?self.upper,
+                ?prev_epoch,
+                ?current_epoch,
+                "fencing previous catalogs"
+            );
+            self.epoch = current_epoch;
+
+            if matches!(self.mode, Mode::Writable) {
+                match self.compare_and_append(fence_updates).await {
+                    Ok(_) => break,
+                    Err(CompareAndAppendError::Fence(e)) => return Err(e.into()),
+                    Err(CompareAndAppendError::UpperMismatch { .. }) => continue,
```

Comment:
Can you add a `tracing::warn!` here so we can more easily identify when this occurs?